cmd/kubeadm/app/cmd/phases/reset/cleanupnode.go

```diff
@@ -128,18 +134,47 @@ func runCleanupNode(c workflow.RunData) error {
 	return nil
 }
 
-func absoluteKubeletRunDirectory() (string, error) {
+// cleanKubeletRunDir ...
+func cleanKubeletRunDir(runDir string, retainDirs []string) error {
+	// If kubelet run dir doesn't even exist there's nothing todo, and we are not consider this an error
+	if _, err := os.Stat(runDir); os.IsNotExist(err) {
+		return nil
+	}
+
+	err := filepath.WalkDir(runDir, func(path string, d os.DirEntry, err error) error {
+		if err != nil {
+			if os.IsNotExist(err) {
+				err = nil // parent was deleted
+			}
+			return err
+		}
+		isRetainPath, isSkipPath := checkPath(path, retainDirs)
+		if isRetainPath {
+			return nil
+		}
```

Comment:
Emm... We still need to be careful with this change, which will cause kubeadm not to be able to fully clean kubelet dir. Previously, the data files would be wiped clean, either by `unmount` or by`delete/remove files`, this guarantees that when the user tries the join the node again, it will almost always be successful. But when this change is added, If unmount fails, the data is retained, and some dirty data(such as wrong cpu_manager_state file?)  may cause kubelet to fail to start, causing confusion for our users. 

Is `kubeamd reset` responsible for data retention? Shouldn't it try to do a thorough cleaning? I have some hesitation as to whether backup data should be left to the users themselves.

cc @neolit123 @pacoxu

Reply:
i've seen a number of complains that deleting unmounted paths is bad. that is true for reset today.

perhaps the -f should matter here.
-f would means full cleanup.

Reply:
@mowangdk when does this problem happen and how often? when is an unmount unable to compete, resulting in kubeadm skipping it and just deleting a path.

i don't think i've seen this on my side...

if a mount path was deleted before unmount finished, how can the user cleanup this mount manually?

@SataQiu 's concerns are valid. this is a change in behavior for reset...if we add a change in behavior, ideally it must be opt in, e g. optional.

Reply:
> ideally it must be opt in, e g. optional.

+1 for optional

Reply:
problem is that we don't want to change the behavior of -f either. maybe a new flag will be needed that 1) forces unmounts, 2) errors if unmount fails

cc @chendave 

@mowangdk please put the work on hold for now and answer the questions here:
https://github.com/kubernetes/kubernetes/pull/120377#discussion_r1328305570

as i said, i few times already we need to plan before writing the code.

Reply:
> @mowangdk when does this problem happen and how often? when is an unmount unable to compete, resulting in kubeadm skipping it and just deleting a path.
> 
> i don't think i've seen this on my side...
> 
> if a mount path was deleted before unmount finished, how can the user cleanup this mount manually?
> 
> @SataQiu 's concerns are valid. this is a change in behavior for reset...if we add a change in behavior, ideally it must be opt in, e g. optional.

Okay, I'll explain the situation in detail.

The cause was an instance of data deletion found on our cloud. So we started troubleshooting the issue. The user scenario is to remove a node from the cluster without draining it, and then realize that the data in the storage previously mounted on the node has been lost. According to the audit logs of the storage,  the ```os.RemoveAll``` command was triggered after removing the node to remove all data from the mount point directory.

So we started troubleshooting what operations on the node would cause the data to be deleted. We found ```kubeadm reset -f```, and found the following in the kubeadm logs 

```
[preflight] Running pre-flight checks
[reset] No etcd config found. Assuming external etcd
[reset] Please, manually reset etcd to prevent further issues
[reset] Stopping the kubelet service
[reset] Unmounting mounted directories in "/var/lib/kubelet"
[removeetcdmember.go:80] [reset] No kubeadm config, using etcd pod spec to get data directory
[unmount_linux.go:48] [reset] Failed to unmount mounted directory in /var/lib/kubelet/: /var/lib/kubelet/pods/xxxx/volumes/kubernetes.io~csi/{pv-name}/mount
[reset] Deleting contents of config directories: [/etc/kubernetes/manifests /etc/kubernetes/pki]
[reset] Deleting files: [/etc/kubernetes/admin.conf /etc/kubernetes/kubelet.conf /etc/kubernetes/bootstrap-kubelet.conf /etc/kubernetes/controller-manager.conf /etc/kubernetes/scheduler.conf]
[reset] Deleting contents of stateful directories: [/var/lib/kubelet /var/lib/dockershim /var/run/kubernetes /var/lib/cni]
```
kubeadm fails in its attempt to clean up the mount point, and starts cleaning up the kubelet's root directory without any special logic after reporting the error. 

IMHO, Whether or not the -f argument is used, removing the contents of a user's mount point does not work as expected. and this problem is present whenever a mount point removal fails.

The reproduce process is very simple, just go to the storage mount directory on the node and trigger ```kubeadm reset -f``` in another terminal. You'll see that the data in the mount point you entered under the kubelet will be cleaned up

Reply:
thanks for the details.

> [reset] Deleting contents of stateful directories: [/var/lib/kubelet /var/lib/dockershim /var/run/kubernetes /var/lib/cni]
kubeadm fails in its attempt to clean up the mount point, and starts cleaning up the kubelet's root directory without any special logic after reporting the error.

which is by design - `kubeadm reset` is a best effort command. it may or may not fully cleanup a node, including unmounting paths or stopping containers.

> [unmount_linux.go:48] [reset] Failed to unmount mounted directory in /var/lib/kubelet/: /var/lib/kubelet/pods/xxxx/volumes/kubernetes.io~csi/{pv-name}/mount

do you know why this unmount failed exactly? i.e. in Linux terminology?

> kubeadm fails in its attempt to clean up the mount point, and starts cleaning up the kubelet's root directory without any special logic after reporting the error.

in this specific scenario, as an administrator, what actions would you take if `kubeadm reset -f` reported the unmount error instead of silently failing and then deleting the whole kubelet directory?

Reply:
opt in with a new reset option seems best for such cases. @chendave what are your thoughts on this?

Reply:
I do think we should give user a choice before wiping it out in this case, so I am +1 to have a option.

I am thinking is it possible to take it this way, "-f" mean best-effort while stopping to clean them without the flag, properly not.

anyway, a config here is helpful and can address the problem.

Reply:
> I am thinking is it possible to take it this way, "-f" mean best-effort while stopping to clean them without the flag, properly not.

-f was originally added to bypass the yes/no prompt. so if we bind this change to -f, users will not be able to both skip the prompt and receive the error (not do the best effort cleanup).

Reply:
@neolit123 Is there a conclusion here?

Reply:
sounds like we have to do the following:
- add new ResetConfiguration field - waitForContainerUnmount
- exact name must be decided
- default value should be false. if false it behaves the way it is today. if true it tries thoroughly to unmount and errors on problems (the way you want).
- we must not add a CLI flag, config is preferred and we are moving away from flags

@SataQiu @chendave WDYT?

Reply:
sure~

Reply:
I am +1 for a new configuration, looks like this is more flexible.